# Exercise 2 - shell, pipes, and csvkit

In this exercise, we'll review a few shell commands and explore working with pipes and csvkit.

You will need to fill in commands whenever prompted.  Please replace the text with your solution.

Remember to submit your completed `.ipynb` file to Blackboard and to add/commit it to your Git repository and push it to GitHub.


## Part 1 - shell commands, redirection, and pipes

### Basic shell commands and redirection

Create a directory called `part1` using `mkdir`.

In [1]:
!mkdir part1

Rename `part1` to `partone` using `mv`.

In [2]:
!mv part1 partone
!ls

exercise-01.ipynb exercise-02.ipynb partone


Create a file named `filelist.txt` using the output from `ls` and the output redirector `>`.

In [3]:
!ls > filelist.txt

In [4]:
!cat filelist.txt

exercise-01.ipynb
exercise-02.ipynb
filelist.txt
partone


Append to `filelist.txt` using the output appending redirector `>>`.  Note the difference between the single `>` and double `>>`.

In [5]:
!ls >> filelist.txt
!cat filelist.txt

exercise-01.ipynb
exercise-02.ipynb
filelist.txt
partone
exercise-01.ipynb
exercise-02.ipynb
filelist.txt
partone


In [6]:
!ls > filelist.txt
!cat filelist.txt

exercise-01.ipynb
exercise-02.ipynb
filelist.txt
partone


What's the difference between `>` and `>>`?


Both >> and > redirect outputs of a command to a file. If the file does not exists in the path, they create the file. Otherwise,  > overwrites the file. On the other hand, >> appends the outputs to the end of the file.

### Your turn

Complete the following tasks in the cells provided.  All the tests in the testing cells (with the `assert` statements) below should pass without error - be sure to execute those as well, and if you see errors, fix your answer and try testing again until there are no errors.

Create a directory called `mydirectory`.

In [7]:
# YOUR CODE HERE
!mkdir mydirectory 

In [8]:
import os
assert 'mydirectory' in os.listdir('.')

Using `ls` and output redirection, create a file called `myfiles.txt` in the directory `mydirectory` that contains the list of files in the current directory.

In [9]:
# YOUR CODE HERE
!ls > ./mydirectory/myfiles.txt

In [10]:
assert 'myfiles.txt' in os.listdir('mydirectory')

In [11]:
myfiles = open('mydirectory/myfiles.txt').read()
assert 'exercise-02.ipynb' in myfiles

Clean up the directory you just created by removing its contents (the file you created) using `rm`.

In [12]:
!rm ./mydirectory/myfiles.txt

In [13]:
assert 'myfiles.txt' not in os.listdir('mydirectory')

Now remove the directory itself using `rmdir`.

In [14]:
# YOUR CODE HERE
!rmdir mydirectory/

In [15]:
assert 'mydirectory' not in os.listdir('.')

### Filters and pipes

Let's look at something a little more interesting.  Download the text of Herman Hesse's *Siddhartha* from [Project Gutenberg](http://www.gutenberg.org/):

In [16]:
!wget https://github.com/gwsb-istm-6212-fall-2016/syllabus-and-schedule/raw/master/exercises/pg2500.txt

--2016-09-07 09:51:25--  https://github.com/gwsb-istm-6212-fall-2016/syllabus-and-schedule/raw/master/exercises/pg2500.txt
Resolving github.com... 192.30.253.112
Connecting to github.com|192.30.253.112|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://raw.githubusercontent.com/gwsb-istm-6212-fall-2016/syllabus-and-schedule/master/exercises/pg2500.txt [following]
--2016-09-07 09:51:26--  https://raw.githubusercontent.com/gwsb-istm-6212-fall-2016/syllabus-and-schedule/master/exercises/pg2500.txt
Resolving raw.githubusercontent.com... 151.101.32.133
Connecting to raw.githubusercontent.com|151.101.32.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 241176 (236K) [text/plain]
Saving to: ‘pg2500.txt’

pg2500.txt          100%[===================>] 235.52K  1.42MB/s    in 0.2s    

2016-09-07 09:51:27 (1.42 MB/s) - ‘pg2500.txt’ saved [241176/241176]



*Note*: sometimes Project Gutenberg restricts access.  If that creates an error for you, you should be able to `wget` the same file from our class repository on GitHub at the url https://github.com/gwsb-istm-6212-fall-2016/syllabus-and-schedule/raw/master/exercises/pg2500.txt.

However you get the file, let's rename it to something easier to remember.

In [17]:
!mv pg2500.txt siddhartha.txt

`head` and `tail` are very useful.  They let you take a quick peek at the start and end of files.

In [18]:
!head siddhartha.txt

In [19]:
!tail siddhartha.txt

`grep` is one of the most useful filters.  It lets you search for and match lines that contain specific expressions.  For example, to find mentions of "copyright":

In [20]:
!grep copyright siddhartha.txt

Notice anything that those lines have in common?

Let's add a little more information by including the `-n` flag to add matching line numbers.

In [21]:
!grep -n copyright siddhartha.txt

Now let's look for any mention of "river".  This will match a lot of text, so we'll just take the first 10 matching lines by *piping* the output from `grep` into `head`.

In [22]:
!grep -n river siddhartha.txt | head

How many lines contain "river"?  We can count by piping into the word count tool `wc`.

In [23]:
!grep river siddhartha.txt | wc

     109    1365    7443


That's 109 matching lines, containing 1365 words and 7979 characters.  If you just wanted the lines by themselves, use `wc -l`:

In [24]:
!grep river siddhartha.txt | wc -l

     109


What if we want to match both upper- and lower-case text?  Use `grep -i`:

In [25]:
!grep time siddhartha.txt | wc -l

     166


In [26]:
!grep -i time siddhartha.txt | wc -l

     167


### Your turn

How many lines in *Siddhartha* contain "other" (just lower-case)?  Start by using `grep` to extract lines that match the word "other" in `siddhartha.txt` and redirecting it to a file called `other-lines.txt`.

In [27]:
# YOUR CODE HERE
!grep other siddhartha.txt > other-lines.txt

In [28]:
%sc h_other = head -1 other-lines.txt
assert "other" in h_other

In [29]:
%sc t_other = tail -1 other-lines.txt
assert "other" in t_other

Now count up the lines in the file you created using `wc`.

In [32]:
# it seems _ does not hold outputs of last executed statement in the notebook. 
# I use the shell capture magic here.
%sc  _ = wc -l other-lines.txt

In [35]:
assert "127" in _

Your answer should be 127!

### Counting words with `grep`

By piping commands together we can do a lot of powerful things right at the command line.  Let's create a count of the most commonly occurring words in *Siddhartha*.  To do that, we could write a Python script that just counts words, but with the command line shell tools we only need to put a proper pipeline together and we can often accomplish tasks like this in one line.

First we need to split up the text lines into a word per line.  There are `grep` flags for that!

In [40]:
!cat siddhartha.txt | grep -oE '\w{{2,}}' | head -10

The
Project
Gutenberg
EBook
of
Siddhartha
by
Herman
Hesse
This


Now we need to sort them and count the unique tokens.  `sort` solves the first problem.

In [48]:
!cat siddhartha.txt | grep -oE '\w{{2,}}' | sort | head -10

000
1500
1887
20
2001
2008
2011
2013
23
2500
sort: write failed: standard output: Broken pipe
sort: write error


And `uniq -c` solves the second problem.

In [42]:
!cat siddhartha.txt | grep -oE '\w{{2,}}' | sort | uniq -c | head -25

   1 000
   1 1500
   1 1887
   1 20
   1 2001
   1 2008
   1 2011
   1 2013
   1 23
   4 2500
   1 30
   1 4557
   1 50
   2 501
   1 596
   1 60
   1 6221541
   1 64
   1 801
   1 809
   1 84116
   2 90
   1 99712
   1 ACTUAL
   2 AGREE


But there's a catch... do you see it?

We need to convert all the words down into lower case so that we are correctly counting unique words.  There's another command, `tr`, for that.

In [43]:
!cat siddhartha.txt | grep -oE '\w{{2,}}' | tr '[:upper:]' '[:lower:]' | sort | uniq -c | head -25

   1 000
   1 1500
   1 1887
   1 20
   1 2001
   1 2008
   1 2011
   1 2013
   1 23
   4 2500
   1 30
   1 4557
   1 50
   2 501
   1 596
   1 60
   1 6221541
   1 64
   1 801
   1 809
   1 84116
   2 90
   1 99712
   2 abandon
   2 abandoned


...and if we want to know only the top 10 words in Siddhartha, we need to sort the output.

In [45]:
!cat siddhartha.txt | grep -oE '\w{{2,}}' | tr '[:upper:]' '[:lower:]' | sort | uniq -c | sort | head -10

   1 000
   1 1500
   1 1887
   1 20
   1 2001
   1 2008
   1 2011
   1 2013
   1 23
   1 30


But that sorts by character, not number.  Fortunately, `sort -n` does what we want.

In [46]:
!cat siddhartha.txt | grep -oE '\w{{2,}}' | tr '[:upper:]' '[:lower:]' | sort | uniq -c | sort -n | head -10

   1 000
   1 1500
   1 1887
   1 20
   1 2001
   1 2008
   1 2011
   1 2013
   1 23
   1 30


But that's the wrong end of the list!  Two ways to fix that:  (a) use `tail` instead of `head`; (b) use `sort -rn`, which will sort in reverse order.  Let's try the latter.

In [47]:
!cat siddhartha.txt | grep -oE '\w{{2,}}' | tr '[:upper:]' '[:lower:]' | sort | uniq -c | sort -rn | head -10

2221 the
1434 and
1225 to
1106 of
 960 he
 708 his
 686 in
 540 you
 524 had
 512 was


### Your turn

Download *Alice in Wonderland* from http://www.gutenberg.org/cache/epub/11/pg11.txt (or https://github.com/gwsb-istm-6212-fall-2016/syllabus-and-schedule/raw/master/exercises/pg11.txt if the first url doesn't work).

In [49]:
# YOUR CODE HERE
!wget https://github.com/gwsb-istm-6212-fall-2016/syllabus-and-schedule/raw/master/exercises/pg11.txt 

--2016-09-07 10:40:18--  https://github.com/gwsb-istm-6212-fall-2016/syllabus-and-schedule/raw/master/exercises/pg11.txt
Resolving github.com... 192.30.253.112
Connecting to github.com|192.30.253.112|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://raw.githubusercontent.com/gwsb-istm-6212-fall-2016/syllabus-and-schedule/master/exercises/pg11.txt [following]
--2016-09-07 10:40:18--  https://raw.githubusercontent.com/gwsb-istm-6212-fall-2016/syllabus-and-schedule/master/exercises/pg11.txt
Resolving raw.githubusercontent.com... 151.101.32.133
Connecting to raw.githubusercontent.com|151.101.32.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 167518 (164K) [text/plain]
Saving to: ‘pg11.txt’

pg11.txt            100%[===================>] 163.59K  --.-KB/s    in 0.05s   

2016-09-07 10:40:20 (3.21 MB/s) - ‘pg11.txt’ saved [167518/167518]



In [52]:
assert 'pg11.txt' in os.listdir('.')

Now rename `pg11.txt` to `alice.txt`.

In [54]:
# YOUR CODE HERE
!mv pg11.txt alice.txt

In [58]:
assert 'alice.txt' in os.listdir('.')

In [60]:
!cat siddhartha.txt | grep -oE '\w{{2,}}' 


The
Project
Gutenberg
EBook
of
Siddhartha
by
Herman
Hesse
This
eBook
is
for
the
use
of
anyone
anywhere
at
no
cost
and
with
almost
no
restrictions
whatsoever
You
may
copy
it
give
it
away
or
re
use
it
under
the
terms
of
the
Project
Gutenberg
License
included
with
this
eBook
or
online
at
www
gutenberg
org
Title
Siddhartha
Author
Herman
Hesse
Translator
Gunther
Olesch
Anke
Dreher
Amy
Coulter
Stefan
Langer
and
Semyon
Chaichenets
Release
Date
April
2008
EBook
2500
Last
updated
July
2011
Last
updated
January
23
2013
Language
English
START
OF
THIS
PROJECT
GUTENBERG
EBOOK
SIDDHARTHA
Produced
by
Michael
Pullen
Chandra
Yenco
Isaac
Jones
SIDDHARTHA
An
Indian
Tale
by
Hermann
Hesse
FIRST
PART
To
Romain
Rolland
my
dear
friend
THE
SON
OF
THE
BRAHMAN
In
the
shade
of
the
house
in
the
sunshine
of
the
riverbank
near
the
boats
in
the
shade
of
the
Sal
wood
forest
in
the
shade
of
the
fig
tree
is
where
Siddhartha
grew
up
the
handsome
son
of
the
Brahman
the
young
falcon
together
with
his
friend
Govinda
son
of


Take a look at the next cell.  Will it find the top 25 unique words in *Alice in Wonderland* successfully?

In [62]:
!cat alice.txt | grep -oE '\w{{2,}}' | sort | uniq -c | head -25

   1 000
   4 11
   1 1500
   1 1887
   1 1994
   2 20
   1 2001
   1 2008
   1 2011
   1 25
   1 30
   1 4557
   1 50
   2 501
   1 596
   1 60
   1 6221541
   1 64
   1 801
   1 809
   1 84116
   2 90
   1 99712
   1 ACTUAL
   3 ADVENTURES


Describe what needs to be done to the previous cell to get it to work correctly.  Describe it using words, explaining the issues, rather than using shell commands!

We have to sort reversely the outputs of "uniq -c" by the count of words before using "head -25" to retrieve the top 25 unique words. And We have to transform uppercase letters to lowercase after the grep command. We want to put the transform command before the first sort command because we want to reduce number of words that the sort command has to process. 

Okay, now implement your solution using shell commands with a pipeline.

In [63]:
# YOUR CODE HERE
!cat alice.txt | grep -oE '\w{{2,}}' | tr '[:upper:]' '[:lower:]' | sort | uniq -c | sort -rn | head -25

1818 the
 940 and
 809 to
 631 of
 610 it
 553 she
 481 you
 462 said
 431 in
 403 alice
 358 was
 330 that
 274 as
 248 her
 228 with
 227 at
 204 on
 200 all
 181 this
 179 for
 178 had
 175 but
 167 be
 166 not
 155 they


## Part 2 - csvkit basics

Let's look at some CSV data using csvkit.  Download the 2015 fourth quarter trip dataset from [Capital Bikeshare's trip history data](https://www.capitalbikeshare.com/trip-history-data):

In [64]:
!wget https://s3.amazonaws.com/capitalbikeshare-data/2015-Q4-cabi-trip-history-data.zip

--2016-09-07 11:11:20--  https://s3.amazonaws.com/capitalbikeshare-data/2015-Q4-cabi-trip-history-data.zip
Resolving s3.amazonaws.com... 52.216.225.107
Connecting to s3.amazonaws.com|52.216.225.107|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 14308794 (14M) [application/zip]
Saving to: ‘2015-Q4-cabi-trip-history-data.zip’

2015-Q4-cabi-trip-h 100%[===================>]  13.65M  3.76MB/s    in 3.8s    

2016-09-07 11:11:25 (3.63 MB/s) - ‘2015-Q4-cabi-trip-history-data.zip’ saved [14308794/14308794]



Let's unzip it, rename it to something short, and take a look.

In [65]:
!unzip 2015-Q4-cabi-trip-history-data.zip

Archive:  2015-Q4-cabi-trip-history-data.zip
  inflating: 2015-Q4-Trips-History-Data.csv  


In [66]:
!mv 2015-Q4-Trips-History-Data.csv q4.csv

In [67]:
!head q4.csv

csvkit gives us great tools for examining and working with CSV data.  We start by looking at the columns:

In [68]:
!csvcut -n q4.csv

  1: Duration (ms)
  2: Start date
  3: End date
  4: Start station number
  5: Start station
  6: End station number
  7: End station
  8: Bike #
  9: Member type


We can also extract just a few columns with `csvcut`:

In [69]:
!csvcut -c1,5,7 q4.csv | head -10

Duration (ms),Start station,End station
166050,Park Rd & Holmead Pl NW,14th & Harvard St NW
379172,34th & Water St NW,25th St & Pennsylvania Ave NW
696038,17th & Corcoran St NW,17th & Corcoran St NW
219423,Adams Mill & Columbia Rd NW,Calvert St & Woodley Pl NW
253230,11th & Kenyon St NW,11th & Kenyon St NW
655251,18th St & Pennsylvania Ave NW,18th St & Wyoming Ave NW
309212,11th & S St NW,18th & R St NW
776195,34th St & Wisconsin Ave NW,17th & Corcoran St NW
151604,Wilson Blvd & N Uhle St,N Veitch  & 20th St N


...and make it look better with `csvlook`:

In [70]:
!csvcut -c1,5,7 q4.csv | head -10 | csvlook

|----------------+-------------------------------+--------------------------------|
|  Duration (ms) | Start station                 | End station                    |
|----------------+-------------------------------+--------------------------------|
|  166050        | Park Rd & Holmead Pl NW       | 14th & Harvard St NW           |
|  379172        | 34th & Water St NW            | 25th St & Pennsylvania Ave NW  |
|  696038        | 17th & Corcoran St NW         | 17th & Corcoran St NW          |
|  219423        | Adams Mill & Columbia Rd NW   | Calvert St & Woodley Pl NW     |
|  253230        | 11th & Kenyon St NW           | 11th & Kenyon St NW            |
|  655251        | 18th St & Pennsylvania Ave NW | 18th St & Wyoming Ave NW       |
|  309212        | 11th & S St NW                | 18th & R St NW                 |
|  776195        | 34th St & Wisconsin Ave NW    | 17th & Corcoran St NW          |
|  151604        | Wilson Blvd & N Uhle St       | N Veitch  & 20th St N    

It gets even better.  Try `csvgrep`:

In [71]:
!csvcut -c1,5,7 q4.csv | csvgrep -c3 -m '21st & I St NW' | head -10 | csvlook

|----------------+----------------------------------+-----------------|
|  Duration (ms) | Start station                    | End station     |
|----------------+----------------------------------+-----------------|
|  872420        | 5th St & Massachusetts Ave NW    | 21st & I St NW  |
|  1249207       | 11th & Kenyon St NW              | 21st & I St NW  |
|  548066        | 17th & Corcoran St NW            | 21st & I St NW  |
|  604362        | 18th St & Wyoming Ave NW         | 21st & I St NW  |
|  1556978       | Army Navy Dr & S Nash St         | 21st & I St NW  |
|  586738        | New Hampshire Ave & T St NW      | 21st & I St NW  |
|  259421        | 17th & K St NW / Farragut Square | 21st & I St NW  |
|  269179        | M St & Pennsylvania Ave NW       | 21st & I St NW  |
|  227962        | 25th St & Pennsylvania Ave NW    | 21st & I St NW  |
|----------------+----------------------------------+-----------------|


But wait, there's more:

In [72]:
!csvcut -c1,5,7 q4.csv | csvgrep -c3 -m '21st & I St NW' | csvsort -c2 | head -10 | csvlook

|----------------+-----------------------+-----------------|
|  Duration (ms) | Start station         | End station     |
|----------------+-----------------------+-----------------|
|  890758        | 10th & E St NW        | 21st & I St NW  |
|  1074547       | 10th & E St NW        | 21st & I St NW  |
|  751900        | 10th & E St NW        | 21st & I St NW  |
|  938747        | 10th & Florida Ave NW | 21st & I St NW  |
|  1001635       | 10th & Florida Ave NW | 21st & I St NW  |
|  837341        | 10th & Florida Ave NW | 21st & I St NW  |
|  891224        | 10th & Florida Ave NW | 21st & I St NW  |
|  1070924       | 10th & Florida Ave NW | 21st & I St NW  |
|  799070        | 10th & Florida Ave NW | 21st & I St NW  |
|----------------+-----------------------+-----------------|


And you can perform basic statistics very easily:

In [73]:
!csvcut -c1,5,7 q4.csv | csvgrep -c3 -m '21st & I St NW' | csvcut -c1 | csvstat

  1. Duration (ms)
	<class 'int'>
	Nulls: False
	Min: 65972
	Max: 29254066
	Sum: 3721091088
	Mean: 867387.2
	Median: 678586.5
	Standard Deviation: 1148902.3949097062
	Unique values: 4281
	5 most frequent values:
		889715:	2
		948623:	2
		716896:	2
		134297:	2
		586081:	2

Row count: 4290


### Your turn

Which set of trips had the longer average trip duration:  trips *starting* at "Massachusetts Ave & Dupont Circle NW", or trips *ending* at "Massachusetts Ave & Dupont Circle NW"?

Use as many new cells as you need to compute the answer, and then write in your answer below in the "YOUR ANSWER HERE" cell.

In [74]:
# the average duration of trips starting at "Massachusetts Ave & Dupont Circle NW"
!csvcut -c1,5 q4.csv | csvgrep -c2 -m "Massachusetts Ave & Dupont Circle NW" | csvcut -c1 | csvstat | grep Mean

	Mean: 946431.0750804995


In [75]:
# the average duration of trips ending at "Massachusetts Ave & Dupont Circle NW"
!csvcut -c1,7 q4.csv | csvgrep -c2 -m "Massachusetts Ave & Dupont Circle NW" | csvcut -c1 | csvstat | grep Mean

	Mean: 840543.3018618869


Trips starting at "Massachusetts Ave & Dupont Circle NW" had the longer average trip duration than  trips ending at "Massachusetts Ave & Dupont Circle NW". 